### PR TITLE
Fromager 361: Requsts Session Object

### PR DESCRIPTION
**Pull request by rd4398**
Fixes: https://github.com/python-wheel-build/fromager/issues/361

### CODE REVIEW

**Review by shubhbapna — COMMENTED** [review]
Instead of passing it in context, should we create a module instead that exposes this session object. Then instead of passing the context everywhere, we just import the session object?

**Comment by rd4398** [comment]
> Instead of passing it in context, should we create a module instead that exposes this session object. Then instead of passing the context everywhere, we just import the session object?

I had tried separate module few weeks ago in https://github.com/python-wheel-build/fromager/pull/228
The suggestions made there is why I am doing this in context instead of separate module

**Comment by shubhbapna** [comment]
> I had tried separate module few weeks ago in #228
> The suggestions made there is why I am doing this in context instead of separate module

Oh is this the module: https://github.com/python-wheel-build/fromager/pull/228/commits/4cc20c85a5e30e21ac02f4eb3bc7771ff5faf0ba ? The module in this commit defines a class but there is no object that can be used directly on importing. I was thinking something along these lines:

`request_session.py`:
```python
import request
session = request.session()
```

Then say in `resolver.py`:
```python
from .request_session import session

# the function signature doesn't change at all
def get_project_from_pypi(
    project: str,
    extras: typing.Iterable[str],
    sdist_server_url: str,
) -> typing.Iterable[Candidate]:
    """Return candidates created from the project name and extras."""
    simple_index_url = sdist_server_url.rstrip("/") + "/" + project + "/"
    logger.debug("%s: getting available versions from %s", project, simple_index_url)
    # using the session object directly
    data = session.get(simple_index_url).content
```

**Comment by rd4398** [comment]
> > I had tried separate module few weeks ago in #228
> > The suggestions made there is why I am doing this in context instead of separate module
> 
> Oh is this the module: [4cc20c8](https://github.com/python-wheel-build/fromager/commit/4cc20c85a5e30e21ac02f4eb3bc7771ff5faf0ba) ? The module in this commit defines a class but there is no object that can be used directly on importing. I was thinking something along these lines:
> 
> `request_session.py`:
> 
> ```python
> import request
> session = request.session()
> ```
> 
> Then say in `resolver.py`:
> 
> ```python
> from .request_session import session
> 
> # the function signature doesn't change at all
> def get_project_from_pypi(
>     project: str,
>     extras: typing.Iterable[str],
>     sdist_server_url: str,
> ) -> typing.Iterable[Candidate]:
>     """Return candidates created from the project name and extras."""
>     simple_index_url = sdist_server_url.rstrip("/") + "/" + project + "/"
>     logger.debug("%s: getting available versions from %s", project, simple_index_url)
>     # using the session object directly
>     data = session.get(simple_index_url).content
> ```

yeah, that's the one.
Hmm okay, I can do that

**Comment by rd4398** [comment]
> > > I had tried separate module few weeks ago in #228
> > > The suggestions made there is why I am doing this in context instead of separate module
> > 
> > 
> > Oh is this the module: [4cc20c8](https://github.com/python-wheel-build/fromager/commit/4cc20c85a5e30e21ac02f4eb3bc7771ff5faf0ba) ? The module in this commit defines a class but there is no object that can be used directly on importing. I was thinking something along these lines:
> > `request_session.py`:
> > ```python
> > import request
> > session = request.session()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Then say in `resolver.py`:
> > ```python
> > from .request_session import session
> > 
> > # the function signature doesn't change at all
> > def get_project_from_pypi(
> >     project: str,
> >     extras: typing.Iterable[str],
> >     sdist_server_url: str,
> > ) -> typing.Iterable[Candidate]:
> >     """Return candidates created from the project name and extras."""
> >     simple_index_url = sdist_server_url.rstrip("/") + "/" + project + "/"
> >     logger.debug("%s: getting available versions from %s", project, simple_index_url)
> >     # using the session object directly
> >     data = session.get(simple_index_url).content
> > ```
> 
> yeah, that's the one. Hmm okay, I can do that

Done

File: src/fromager/context.py [hunk 1]
```diff
@@ -87,6 +87,9 @@ def __init__(
             set()
         )
 
+        # create a requests session
+        # self.requests = self._make_requests_session()
+
```
Comment (dhellmann) [inline src/fromager/context.py:92]: Leftovers?

File: src/fromager/request_session.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import requests
+from requests.adapters import HTTPAdapter
+
+session = requests.Session()
+adapter = HTTPAdapter(max_retries=5)
```
Comment (dhellmann) [inline src/fromager/request_session.py:5]: Does adapter need to be public, or could it be renamed `_adapter`?

**Review by dhellmann — COMMENTED** [review]
A couple of nits, but this generally looks good.

File: src/fromager/context.py (hunk 1 above)
Comment (rd4398) [reply src/fromager/context.py]: yeah, removed just now!

File: src/fromager/request_session.py (hunk 2 above)
Comment (rd4398) [reply src/fromager/request_session.py]: yeah, renamed to `_adapter`

**Comment by rd4398** [comment]
> I'm happy with this version. @shubhbapna had comments, so I'll leave it open for him to review.
> 
> I re-ran the coverage job that failed and it failed again. Is that a known problem?

I am not sure. I am seeing it for the first time

**Review by shubhbapna — APPROVED** [review]
LGTM! I reran the checks in hopes of regenerating the coverage files. The coverage job was failing because no artifacts were found for some reason

**Pull request merged**